### ISSUE TITLE
BatchNorm is broken

### ISSUE THREAD

**Issue by Roseck16** [FluxML/Flux.jl#976]
The layer batch norm doesn't work, even with it's own example:

```
m = Chain(
  Dense(28^2, 64),
  BatchNorm(64, relu),
  Dense(64, 10),
  BatchNorm(10),
  softmax)

julia> m(rand(28^2))
ERROR: arraysize: dimension out of range
Stacktrace:
 [1] size at .\array.jl:154 [inlined]
 [2] (::BatchNorm{typeof(relu),Array{Float32,1},Array{Float32,1},Float32})(::Array{Float32,1}) at C:\Users\user\.julia\packages\Flux\oX9Pi\src\layers\normalise.jl:140
 [3] applychain at C:\Users\user\.julia\packages\Flux\oX9Pi\src\layers\basic.jl:30 [inlined] (repeats 2 times)
 [4] (::Chain{Tuple{Dense{typeof(identity),Array{Float32,2},Array{Float32,1}},BatchNorm{typeof(relu),Array{Float32,1},Array{Float32,1},Float32},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}},BatchNorm{typeof(identity),Array{Float32,1},Array{Float32,1},Float32},typeof(softmax)}})(::Array{Float64,1}) at C:\Users\user\.julia\packages\Flux\oX9Pi\src\layers\basic.jl:32
 [5] top-level scope at none:0
```

Using Flux v0.10.0 in Julia 1.2.0

**Comment by chinglamchoi** [comment]
Batch norm currently takes input with at least N=2 dimensions, and calls the N-1'th dimension. So the dimension error comes from calling dim 0 in Julia, which as you know is 1-indexed. So instead of passing `rand(784)`  (size is `(784,) `), we can pass `rand(784, 1)`, which works for me in Flux v0.10.1

For more details please refer to: https://github.com/FluxML/Flux.jl/blob/ddc2c20e68919faa41a04ed39ed3cfa08d6d5189/src/layers/normalise.jl#L101

**Comment by CarloLucibello** [comment]
closing since as @chinglamchoi explained this is not really an issue

**Comment by CarloLucibello** [comment]
Feel free to reopen in case we want to support inputs without batch dimension
